// xero-node/lib/internals/__tests__/config-helper.tests.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var config_helper_1 = require("../config-helper");
var privateKey_helpers_1 = require("./helpers/privateKey-helpers");
var version = require('../../../package.json').version;
describe('config-helper', function () {
    describe('Private apps', function () {
        describe('with cert as a path', function () {
            var xeroConfigWithCertPath = {
                appType: 'private',
                consumerKey: 'RDGDV41TRLQZDFSDX96TKQ2KRJIW4C',
                consumerSecret: 'DJ3CMGDB0DIIA9DNEEJMRLZG0BWE7Y',
                privateKeyPath: privateKey_helpers_1.validTestCertPath()
            };
            it('maps config correctly', function () {
                var retrievedState = config_helper_1.mapConfig(xeroConfigWithCertPath, {});
                expect(retrievedState).toEqual({
                    accept: 'application/json',
                    userAgent: "NodeJS-XeroAPIClient." + version + ".RDGDV41TRLQZDFSDX96TKQ2KRJIW4C",
                    consumerKey: 'RDGDV41TRLQZDFSDX96TKQ2KRJIW4C',
                    consumerSecret: privateKey_helpers_1.testCertString(),
                    tenantType: null,
                    signatureMethod: 'RSA-SHA1',
                    callbackUrl: null,
                    apiBaseUrl: 'https://api.xero.com',
                    apiBasePath: '',
                    oauthAccessTokenPath: '/oauth/AccessToken',
                    oauthRequestTokenPath: '/oauth/RequestToken',
                });
            });
            it('maps state correctly', function () {
                var retrievedState = config_helper_1.mapState(xeroConfigWithCertPath);
                expect(retrievedState).toEqual({
                    oauth_token: 'RDGDV41TRLQZDFSDX96TKQ2KRJIW4C',
                    oauth_token_secret: privateKey_helpers_1.testCertString()
                });
            });
        });
        describe('with cert as a string', function () {
            var xeroConfigWithCertPath = {
                appType: 'private',
                consumerKey: 'RDGDV41TRLQZDFSDX96TKQ2KRJIW4C',
                consumerSecret: 'DJ3CMGDB0DIIA9DNEEJMRLZG0BWE7Y',
                privateKeyString: privateKey_helpers_1.testCertString()
            };
            it('maps config correctly', function () {
                var retrievedState = config_helper_1.mapConfig(xeroConfigWithCertPath, {});
                expect(retrievedState).toEqual({
                    accept: 'application/json',
                    userAgent: "NodeJS-XeroAPIClient." + version + ".RDGDV41TRLQZDFSDX96TKQ2KRJIW4C",
                    consumerKey: 'RDGDV41TRLQZDFSDX96TKQ2KRJIW4C',
                    consumerSecret: privateKey_helpers_1.testCertString(),
                    tenantType: null,
                    apiBasePath: '',
                    signatureMethod: 'RSA-SHA1',
                    callbackUrl: null,
                    apiBaseUrl: 'https://api.xero.com',
                    oauthAccessTokenPath: '/oauth/AccessToken',
                    oauthRequestTokenPath: '/oauth/RequestToken',
                });
            });
            it('maps state correctly', function () {
                var retrievedState = config_helper_1.mapState(xeroConfigWithCertPath);
                expect(retrievedState).toEqual({
                    oauth_token: 'RDGDV41TRLQZDFSDX96TKQ2KRJIW4C',
                    oauth_token_secret: privateKey_helpers_1.testCertString()
                });
            });
        });
    });
    describe('Public apps', function () {
        var xeroConfig = {
            appType: 'public',
            consumerKey: 'RDGDV41TRLQZDFSDX96TKQ2KRJIW4C',
            consumerSecret: 'DJ3CMGDB0DIIA9DNEEJMRLZG0BWE7Y',
            privateKeyPath: privateKey_helpers_1.validTestCertPath()
        };
        it('maps config correctly', function () {
            var retrievedState = config_helper_1.mapConfig(xeroConfig, {});
            expect(retrievedState).toEqual({
                accept: 'application/json',
                userAgent: "NodeJS-XeroAPIClient." + version + ".RDGDV41TRLQZDFSDX96TKQ2KRJIW4C",
                consumerKey: 'RDGDV41TRLQZDFSDX96TKQ2KRJIW4C',
                consumerSecret: 'DJ3CMGDB0DIIA9DNEEJMRLZG0BWE7Y',
                tenantType: null,
                apiBasePath: '',
                signatureMethod: 'HMAC-SHA1',
                callbackUrl: null,
                apiBaseUrl: 'https://api.xero.com',
                oauthAccessTokenPath: '/oauth/AccessToken',
                oauthRequestTokenPath: '/oauth/RequestToken',
            });
        });
        it('maps state correctly', function () {
            var retrievedState = config_helper_1.mapState(xeroConfig);
            expect(retrievedState).toEqual(null);
        });
    });
    describe('Partner apps', function () {
        describe('with cert as a path', function () {
            var xeroConfig = {
                appType: 'partner',
                consumerKey: 'RDGDV41TRLQZDFSDX96TKQ2KRJIW4C',
                consumerSecret: 'DJ3CMGDB0DIIA9DNEEJMRLZG0BWE7Y',
                privateKeyPath: privateKey_helpers_1.validTestCertPath()
            };
            it('maps config correctly', function () {
                var retrievedState = config_helper_1.mapConfig(xeroConfig, {});
                expect(retrievedState).toEqual({
                    userAgent: "NodeJS-XeroAPIClient." + version + ".RDGDV41TRLQZDFSDX96TKQ2KRJIW4C",
                    accept: 'application/json',
                    consumerKey: 'RDGDV41TRLQZDFSDX96TKQ2KRJIW4C',
                    consumerSecret: privateKey_helpers_1.testCertString(),
                    tenantType: null,
                    apiBasePath: '',
                    signatureMethod: 'RSA-SHA1',
                    callbackUrl: null,
                    apiBaseUrl: 'https://api.xero.com',
                    oauthAccessTokenPath: '/oauth/AccessToken',
                    oauthRequestTokenPath: '/oauth/RequestToken',
                });
            });
            it('maps state correctly', function () {
                var retrievedState = config_helper_1.mapState(xeroConfig);
                expect(retrievedState).toEqual(null);
            });
        });
        describe('with cert as a string', function () {
            var xeroConfig = {
                appType: 'partner',
                consumerKey: 'RDGDV41TRLQZDFSDX96TKQ2KRJIW4C',
                consumerSecret: 'DJ3CMGDB0DIIA9DNEEJMRLZG0BWE7Y',
                privateKeyString: privateKey_helpers_1.testCertString()
            };
            it('maps config correctly', function () {
                var retrievedState = config_helper_1.mapConfig(xeroConfig, {});
                expect(retrievedState).toEqual({
                    accept: 'application/json',
                    userAgent: "NodeJS-XeroAPIClient." + version + ".RDGDV41TRLQZDFSDX96TKQ2KRJIW4C",
                    consumerKey: 'RDGDV41TRLQZDFSDX96TKQ2KRJIW4C',
                    consumerSecret: privateKey_helpers_1.testCertString(),
                    tenantType: null,
                    apiBasePath: '',
                    signatureMethod: 'RSA-SHA1',
                    callbackUrl: null,
                    apiBaseUrl: 'https://api.xero.com',
                    oauthAccessTokenPath: '/oauth/AccessToken',
                    oauthRequestTokenPath: '/oauth/RequestToken',
                });
            });
            it('maps state correctly', function () {
                var retrievedState = config_helper_1.mapState(xeroConfig);
                expect(retrievedState).toEqual(null);
            });
        });
    });
    describe('API Config', function () {
        describe('with cert as path', function () {
            var xeroConfig = {
                appType: 'partner',
                consumerKey: 'RDGDV41TRLQZDFSDX96TKQ2KRJIW4C',
                consumerSecret: 'DJ3CMGDB0DIIA9DNEEJMRLZG0BWE7Y',
                privateKeyPath: privateKey_helpers_1.validTestCertPath()
            };
            var apiConfig = {
                tenantType: 'PRACTICE',
                apiBasePath: '/s/s',
            };
            it('maps config correctly', function () {
                var retrievedState = config_helper_1.mapConfig(xeroConfig, apiConfig);
                expect(retrievedState).toEqual({
                    accept: 'application/json',
                    userAgent: "NodeJS-XeroAPIClient." + version + ".RDGDV41TRLQZDFSDX96TKQ2KRJIW4C",
                    consumerKey: 'RDGDV41TRLQZDFSDX96TKQ2KRJIW4C',
                    consumerSecret: privateKey_helpers_1.testCertString(),
                    tenantType: 'PRACTICE',
                    apiBasePath: '/s/s',
                    signatureMethod: 'RSA-SHA1',
                    callbackUrl: null,
                    apiBaseUrl: 'https://api.xero.com',
                    oauthAccessTokenPath: '/oauth/AccessToken',
                    oauthRequestTokenPath: '/oauth/RequestToken',
                });
            });
        });
        describe('with user agent set', function () {
            var xeroConfig = {
                appType: 'partner',
                userAgent: 'PHILWASHERE',
                consumerKey: 'RDGDV41TRLQZDFSDX96TKQ2KRJIW4C',
                consumerSecret: 'DJ3CMGDB0DIIA9DNEEJMRLZG0BWE7Y',
                privateKeyPath: privateKey_helpers_1.validTestCertPath()
            };
            var apiConfig = {
                tenantType: 'PRACTICE',
                apiBasePath: '/s/s',
            };
            it('appends useragent', function () {
                var retrievedState = config_helper_1.mapConfig(xeroConfig, apiConfig);
                expect(retrievedState).toEqual({
                    accept: 'application/json',
                    userAgent: "NodeJS-XeroAPIClient.PHILWASHERE." + version + ".RDGDV41TRLQZDFSDX96TKQ2KRJIW4C",
                    consumerKey: 'RDGDV41TRLQZDFSDX96TKQ2KRJIW4C',
                    consumerSecret: privateKey_helpers_1.testCertString(),
                    tenantType: 'PRACTICE',
                    apiBasePath: '/s/s',
                    signatureMethod: 'RSA-SHA1',
                    callbackUrl: null,
                    apiBaseUrl: 'https://api.xero.com',
                    oauthAccessTokenPath: '/oauth/AccessToken',
                    oauthRequestTokenPath: '/oauth/RequestToken',
                });
            });
        });
        describe('with cert as string', function () {
            var xeroConfig = {
                appType: 'partner',
                consumerKey: 'RDGDV41TRLQZDFSDX96TKQ2KRJIW4C',
                consumerSecret: 'DJ3CMGDB0DIIA9DNEEJMRLZG0BWE7Y',
                privateKeyString: privateKey_helpers_1.testCertString()
            };
            var apiConfig = {
                tenantType: 'PRACTICE'
            };
            it('maps config correctly', function () {
                var retrievedState = config_helper_1.mapConfig(xeroConfig, apiConfig);
                expect(retrievedState).toEqual({
                    accept: 'application/json',
                    userAgent: "NodeJS-XeroAPIClient." + version + ".RDGDV41TRLQZDFSDX96TKQ2KRJIW4C",
                    consumerKey: 'RDGDV41TRLQZDFSDX96TKQ2KRJIW4C',
                    consumerSecret: privateKey_helpers_1.testCertString(),
                    tenantType: 'PRACTICE',
                    signatureMethod: 'RSA-SHA1',
                    callbackUrl: null,
                    apiBasePath: '',
                    apiBaseUrl: 'https://api.xero.com',
                    oauthAccessTokenPath: '/oauth/AccessToken',
                    oauthRequestTokenPath: '/oauth/RequestToken',
                });
            });
        });
    });
});
