// xero-node/lib/internals/__tests__/error-handling.tests.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var _this = this;
Object.defineProperty(exports, "__esModule", { value: true });
var TestAPIClient_1 = require("./helpers/TestAPIClient");
var privateKey_helpers_1 = require("./helpers/privateKey-helpers");
var XeroError_1 = require("../../XeroError");
var OAuth1HttpClient_1 = require("../OAuth1HttpClient");
var InMemoryOAuthLib_1 = require("./helpers/InMemoryOAuthLib");
describe('HTTP errors', function () {
    var testError = new XeroError_1.XeroError(101, 'the sky is blue', null);
    var inMemoryOAuthFF = new InMemoryOAuthLib_1.InMemoryOAuthLibFactoryFactory();
    var xeroClient;
    var oauthConfig = {
        consumerKey: 'ck',
        consumerSecret: 'cs',
        tenantType: null,
        apiBaseUrl: 'abu',
        apiBasePath: 'abp',
        oauthRequestTokenPath: 'ortp',
        oauthAccessTokenPath: 'oatp',
        signatureMethod: 'sigm',
        callbackUrl: 'http://sdf.sdf',
        accept: 'acceps',
        userAgent: 'ua'
    };
    var requestToken = {
        oauth_token: 'reqtoken',
        oauth_token_secret: 'reqsecret',
    };
    var oauthState = {
        oauth_token: 'atoken',
        oauth_token_secret: 'asecret',
        oauth_session_handle: 'sessionHandle'
    };
    var xeroConfig = {
        appType: 'private',
        consumerKey: 'RDGDV41TRLQZDFSDX96TKQ2KRJIW4C',
        consumerSecret: 'DJ3CMGDB0DIIA9DNEEJMRLZG0BWE7Y',
        privateKeyPath: privateKey_helpers_1.validTestCertPath()
    };
    beforeAll(function () { return __awaiter(_this, void 0, void 0, function () {
        var oAuth1HttpClient;
        return __generator(this, function (_a) {
            oAuth1HttpClient = new OAuth1HttpClient_1.OAuth1HttpClient(oauthConfig, oauthState, inMemoryOAuthFF.newFactory());
            xeroClient = new TestAPIClient_1.TestAPIClient(xeroConfig, null, oAuth1HttpClient);
            return [2 /*return*/];
        });
    }); });
    var fixtures = [
        { verb: 'get', methodUnderTest: function () { return xeroClient.oauth1Client.get('/any-endpoint'); } },
        { verb: 'put', methodUnderTest: function () { return xeroClient.oauth1Client.put('/any-endpoint', { body: 'b' }); } },
        { verb: 'post', methodUnderTest: function () { return xeroClient.oauth1Client.post('/any-endpoint', { body: 'b' }); } },
        { verb: 'delete', methodUnderTest: function () { return xeroClient.oauth1Client.delete('/any-endpoint'); } },
        // TODO { verb: 'writeUTF8ResponseToStream', underlyingMethod: 'get', methodUnderTest: () => xeroClient.oauth1Client.writeUTF8ResponseToStream('/any-endpoint', 'mime/type', null) },
        { verb: 'getUnauthorisedRequestToken', underlyingMethod: 'getOAuthRequestToken', methodUnderTest: function () { return xeroClient.oauth1Client.getRequestToken(); } },
        { verb: 'swapRequestTokenforAccessToken', underlyingMethod: 'getOAuthAccessToken', methodUnderTest: function () { return xeroClient.oauth1Client.swapRequestTokenforAccessToken(requestToken, 'any-token'); } },
        { verb: 'refreshAccessToken', underlyingMethod: '_performSecureRequest', methodUnderTest: function () { return xeroClient.oauth1Client.refreshAccessToken(); } }
    ];
    fixtures.map(function (fixture) {
        describe(fixture.verb, function () {
            beforeEach(function () {
                inMemoryOAuthFF.inMemoryOAuthLib.reset();
                inMemoryOAuthFF.inMemoryOAuthLib.setResponse(true, testError.data, { statusCode: testError.statusCode });
            });
            it('calls correct verb and rejects with XeroError', function () {
                var result = fixture.methodUnderTest();
                expect(result).toBeInstanceOf(Promise);
                expect(result).rejects.toEqual(testError);
                inMemoryOAuthFF.inMemoryOAuthLib.lastCalledThisMethod(fixture.underlyingMethod || fixture.verb);
            });
        });
    });
});
