// xero-node/lib/internals/__tests__/OAuth1HttpClient.tests.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var _this = this;
Object.defineProperty(exports, "__esModule", { value: true });
var InMemoryOAuthLib_1 = require("./helpers/InMemoryOAuthLib");
var OAuth1HttpClient_1 = require("../OAuth1HttpClient");
var requestToken = {
    oauth_token: 'test3',
    oauth_token_secret: 'test4'
};
var defaultState = {
    oauth_token: 'test7',
    oauth_token_secret: 'test8',
    oauth_session_handle: 'test9'
};
describe('OAuth1HttpClient', function () {
    var inMemoryOAuthLib = new InMemoryOAuthLib_1.InMemoryOAuthLibFactoryFactory();
    var oauth1HttpClient;
    var oauthConfig = {
        consumerKey: 'ck',
        consumerSecret: 'cs',
        tenantType: null,
        apiBaseUrl: 'abu',
        apiBasePath: 'abp',
        oauthRequestTokenPath: 'ortp',
        oauthAccessTokenPath: 'oatp',
        signatureMethod: 'RSA-SHA1',
        accept: 'acceps',
        callbackUrl: 'https://fakeurl.com/fake',
        userAgent: 'ua'
    };
    describe('and setting state', function () {
        beforeEach(function () {
            oauth1HttpClient = new OAuth1HttpClient_1.OAuth1HttpClient(oauthConfig, defaultState);
        });
        it('matches what it was set to', function () {
            expect(oauth1HttpClient._state).toEqual(defaultState);
        });
        it('only overrides the accessToken keys', function () {
            var client = oauth1HttpClient;
            client.setState({
                oauth_token: 'something new', oauth_token_secret: 'something borrowed'
            });
            expect(client._state).not.toEqual(defaultState);
            expect(client._state).toEqual({
                oauth_token: 'something new',
                oauth_token_secret: 'something borrowed',
                oauth_session_handle: 'test9'
            });
        });
        it('only overrides the oauth_session_handle keys', function () {
            var client = oauth1HttpClient;
            client.setState({
                oauth_session_handle: 'yoyo'
            });
            expect(client._state).not.toEqual(defaultState);
            expect(client._state).toEqual({
                oauth_token: 'test7',
                oauth_token_secret: 'test8',
                oauth_session_handle: 'yoyo'
            });
        });
    });
    describe('and building authorise url', function () {
        beforeEach(function () { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                oauth1HttpClient = new OAuth1HttpClient_1.OAuth1HttpClient(oauthConfig, defaultState, inMemoryOAuthLib.newFactory());
                return [2 /*return*/];
            });
        }); });
        it('it builds the authorise url', function () {
            expect(oauth1HttpClient.buildAuthoriseUrl(requestToken)).toEqual("abu/oauth/Authorize?oauth_token=test3");
        });
    });
    describe('and getting RequestTokens', function () {
        var token;
        beforeEach(function () { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        oauth1HttpClient = new OAuth1HttpClient_1.OAuth1HttpClient(oauthConfig, defaultState, inMemoryOAuthLib.newFactory());
                        inMemoryOAuthLib.inMemoryOAuthLib.set_getOAuthRequestToken('aaa', 'bbb');
                        return [4 /*yield*/, oauth1HttpClient.getRequestToken()];
                    case 1:
                        token = _a.sent();
                        return [2 /*return*/];
                }
            });
        }); });
        it('token is returned', function () {
            expect(token.oauth_token).toBe('aaa');
            expect(token.oauth_token_secret).toBe('bbb');
        });
    });
    describe('and swapping request for access token', function () {
        var authState;
        beforeAll(function () { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        inMemoryOAuthLib.inMemoryOAuthLib.reset();
                        oauth1HttpClient = new OAuth1HttpClient_1.OAuth1HttpClient(oauthConfig, defaultState, inMemoryOAuthLib.newFactory());
                        inMemoryOAuthLib.inMemoryOAuthLib.set_SwapRequestTokenforAccessToken("access+token", "access+secret", '1800');
                        return [4 /*yield*/, oauth1HttpClient.swapRequestTokenforAccessToken(requestToken, '1234')];
                    case 1:
                        authState = _a.sent();
                        return [2 /*return*/];
                }
            });
        }); });
        it('returns and sets expected state', function () {
            var expectedState = { oauth_token: 'access+token', oauth_token_secret: 'access+secret' };
            var client = oauth1HttpClient;
            expect(authState).toMatchObject(expectedState);
            expect(client._state).toMatchObject(expectedState);
            var timeObject = new Date();
            var expDate = new Date(timeObject.getTime() + (1800 * 1000));
            // Removes seconds from dates...
            expect(client._state.oauth_expires_at.setSeconds(0, 0)).toEqual(expDate.setSeconds(0, 0));
        });
    });
    describe('and refreshing authorized request token', function () {
        var authState;
        beforeAll(function () { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        oauth1HttpClient = new OAuth1HttpClient_1.OAuth1HttpClient(oauthConfig, defaultState, inMemoryOAuthLib.newFactory());
                        inMemoryOAuthLib.inMemoryOAuthLib.set__performSecureRequest("access#token", "access#secret", 'session#handle', '1800');
                        return [4 /*yield*/, oauth1HttpClient.refreshAccessToken()];
                    case 1:
                        authState = _a.sent();
                        return [2 /*return*/];
                }
            });
        }); });
        it('returns sets expected state', function () {
            var currentMilliseconds = new Date().getTime();
            var expDate = new Date(currentMilliseconds + (1800 * 1000));
            var expectedState = {
                oauth_token: "access#token",
                oauth_token_secret: "access#secret",
                oauth_session_handle: 'session#handle',
            };
            expect(authState).toMatchObject(expectedState);
            expect(oauth1HttpClient._state).toMatchObject(expectedState);
            // Removes seconds from dates...
            expect(authState.oauth_expires_at.setSeconds(0, 0)).toEqual(expDate.setSeconds(0, 0));
        });
    });
    describe('and using custom headers', function () {
        var defaultHeaders = {
            'Accept': 'application/json',
            'User-Agent': 'ua'
        };
        describe('on get()', function () {
            it('uses default custom header', function () {
                oauth1HttpClient.get('someEndpoing');
                expect(inMemoryOAuthLib.inMemoryOAuthLib._headers).toEqual(defaultHeaders);
            });
            it('adds a custom headers when passed in', function () {
                oauth1HttpClient.get('someEndpoing', { randomHeader: 'valuehere' });
                expect(inMemoryOAuthLib.inMemoryOAuthLib._headers).toEqual(__assign({}, defaultHeaders, { randomHeader: 'valuehere' }));
            });
            it('then is back to default for next call', function () {
                oauth1HttpClient.get('someEndpoing');
                expect(inMemoryOAuthLib.inMemoryOAuthLib._headers).toEqual(defaultHeaders);
            });
            it('can override default Accept type', function () {
                oauth1HttpClient.get('someEndpoing', { Accept: 'application/pea' });
                expect(inMemoryOAuthLib.inMemoryOAuthLib._headers).toEqual({
                    'Accept': 'application/pea',
                    'User-Agent': 'ua'
                });
            });
        });
        describe('on put()', function () {
            it('uses default custom header', function () {
                oauth1HttpClient.put('someEndpoing', {});
                expect(inMemoryOAuthLib.inMemoryOAuthLib._headers).toEqual(defaultHeaders);
            });
            it('adds a custom headers when passed in', function () {
                oauth1HttpClient.put('someEndpoing', {}, { randomHeader: 'valuehere' });
                expect(inMemoryOAuthLib.inMemoryOAuthLib._headers).toEqual(__assign({}, defaultHeaders, { randomHeader: 'valuehere' }));
            });
            it('then is back to default for next call', function () {
                oauth1HttpClient.put('someEndpoing', {});
                expect(inMemoryOAuthLib.inMemoryOAuthLib._headers).toEqual(defaultHeaders);
            });
            it('can override default Accept type', function () {
                oauth1HttpClient.put('someEndpoing', {}, { Accept: 'application/pea' });
                expect(inMemoryOAuthLib.inMemoryOAuthLib._headers).toEqual({
                    'Accept': 'application/pea',
                    'User-Agent': 'ua'
                });
            });
        });
        describe('on post()', function () {
            it('uses default custom header', function () {
                oauth1HttpClient.post('someEndpoing', {});
                expect(inMemoryOAuthLib.inMemoryOAuthLib._headers).toEqual(defaultHeaders);
            });
            it('adds a custom headers when passed in', function () {
                oauth1HttpClient.post('someEndpoing', {}, { randomHeader: 'valuehere' });
                expect(inMemoryOAuthLib.inMemoryOAuthLib._headers).toEqual(__assign({}, defaultHeaders, { randomHeader: 'valuehere' }));
            });
            it('then is back to default for next call', function () {
                oauth1HttpClient.post('someEndpoing', {});
                expect(inMemoryOAuthLib.inMemoryOAuthLib._headers).toEqual(defaultHeaders);
            });
            it('can override default Accept type', function () {
                oauth1HttpClient.post('someEndpoing', {}, { Accept: 'application/pea' });
                expect(inMemoryOAuthLib.inMemoryOAuthLib._headers).toEqual({
                    'Accept': 'application/pea',
                    'User-Agent': 'ua'
                });
            });
        });
        describe('on delete()', function () {
            it('uses default custom header', function () {
                oauth1HttpClient.delete('someEndpoing');
                expect(inMemoryOAuthLib.inMemoryOAuthLib._headers).toEqual(defaultHeaders);
            });
            it('adds a custom headers when passed in', function () {
                oauth1HttpClient.delete('someEndpoing', { randomHeader: 'valuehere' });
                expect(inMemoryOAuthLib.inMemoryOAuthLib._headers).toEqual(__assign({}, defaultHeaders, { randomHeader: 'valuehere' }));
            });
            it('then is back to default for next call', function () {
                oauth1HttpClient.delete('someEndpoing');
                expect(inMemoryOAuthLib.inMemoryOAuthLib._headers).toEqual(defaultHeaders);
            });
            it('can override default Accept type', function () {
                oauth1HttpClient.delete('someEndpoing', { Accept: 'application/pea' });
                expect(inMemoryOAuthLib.inMemoryOAuthLib._headers).toEqual({
                    'Accept': 'application/pea',
                    'User-Agent': 'ua'
                });
            });
        });
    });
});
