// xero-node/lib/internals/__tests__/helpers/InMemoryOAuthLib.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var InMemoryOAuthLibFactoryFactory = /** @class */ (function () {
    function InMemoryOAuthLibFactoryFactory() {
        this.inMemoryOAuthLib = new InMemoryOAuthLib();
    }
    InMemoryOAuthLibFactoryFactory.prototype.newFactory = function () {
        var _this = this;
        return function (config) {
            _this.inMemoryOAuthLib.setConfig(config);
            return _this.inMemoryOAuthLib;
        };
    };
    return InMemoryOAuthLibFactoryFactory;
}());
exports.InMemoryOAuthLibFactoryFactory = InMemoryOAuthLibFactoryFactory;
var InMemoryOAuthLib = /** @class */ (function () {
    function InMemoryOAuthLib(config) {
        this.config = config;
        this.err = null;
        this.returnData = null;
        this.returnHttpResponse = null;
        this.lastCalledUrl = '';
        this.lastCalledMethod = '';
        this.return_oauth_token = null;
        this.return_oauth_secret = null;
        this.returnAuthorisedToken = null;
        this.returnAuthorisedSecret = null;
        this.returnSessionHandle = null;
        this.lastRequestedBody = null;
        this.oauth_expires_in = null;
    }
    InMemoryOAuthLib.prototype.setConfig = function (config) {
        this.config = config;
        this.config['key'] = 'test';
    };
    InMemoryOAuthLib.prototype.lastCalledThisURL = function (url) {
        expect(this.lastCalledUrl).toBe(url);
    };
    InMemoryOAuthLib.prototype.reset = function () {
        this.err = undefined;
        this.returnData = null;
        this.returnHttpResponse = null;
        this.lastCalledUrl = '';
        this.lastRequestedBody = null;
        this.lastCalledMethod = '';
        this.return_oauth_token = null;
        this.return_oauth_secret = null;
        this.returnAuthorisedToken = null;
        this.returnAuthorisedSecret = null;
        this.returnSessionHandle = null;
        this.oauth_expires_in = null;
    };
    InMemoryOAuthLib.prototype.lastCalledThisMethod = function (verb) {
        expect(this.lastCalledMethod).toBe(verb);
    };
    InMemoryOAuthLib.prototype.set_SwapRequestTokenforAccessToken = function (oauth_token, oauth_secret, oauth_expires_in, sessionHandle) {
        this.returnAuthorisedToken = oauth_token;
        this.returnAuthorisedSecret = oauth_secret;
        this.returnSessionHandle = sessionHandle;
        this.oauth_expires_in = oauth_expires_in;
    };
    InMemoryOAuthLib.prototype.lastRequestedHadBody = function (expectedBody) {
        if (expectedBody) {
            expect(this.lastRequestedBody).toMatch(JSON.stringify(expectedBody));
        }
        else {
            expect(this.lastRequestedBody).toBeNull();
        }
    };
    InMemoryOAuthLib.prototype.lastHadThisHeader = function (expectedHeader) {
        if (expectedHeader) {
            expect(this._headers).toEqual(__assign({
                'Accept': 'application/json',
                'User-Agent': 'NodeJS-XeroAPIClient.RDGDV41TRLQZDFSDX96TKQ2KRJIW4C'
                // tslint:disable-next-line:align
            }, expectedHeader));
        }
    };
    InMemoryOAuthLib.prototype.get = function (url, oauthToken, oauthSecret, callback) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.lastCalledUrl = url;
                this.lastCalledMethod = 'get';
                callback(this.err, this.returnData, this.returnHttpResponse);
                return [2 /*return*/];
            });
        });
    };
    InMemoryOAuthLib.prototype.post = function (url, oauthToken, oauthSecret, body, contentType, callback) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.lastCalledUrl = url;
                this.lastCalledMethod = 'post';
                this.lastRequestedBody = body;
                callback(this.err, this.returnData, this.returnHttpResponse);
                return [2 /*return*/];
            });
        });
    };
    InMemoryOAuthLib.prototype.delete = function (url, oauthToken, oauthSecret, callback) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.lastCalledUrl = url;
                this.lastCalledMethod = 'delete';
                callback(this.err, this.returnData, this.returnHttpResponse);
                return [2 /*return*/];
            });
        });
    };
    InMemoryOAuthLib.prototype.put = function (url, oauthToken, oauthSecret, body, contentType, callback) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.lastCalledUrl = url;
                this.lastRequestedBody = body;
                this.lastCalledMethod = 'put';
                callback(this.err, this.returnData, this.returnHttpResponse);
                return [2 /*return*/];
            });
        });
    };
    InMemoryOAuthLib.prototype.setResponse = function (isErr, returnGetData, returnGetHttpResponse) {
        this.err = isErr ? __assign({}, returnGetHttpResponse, { data: returnGetData }) : undefined;
        this.returnData = returnGetData;
        this.returnHttpResponse = returnGetHttpResponse;
    };
    InMemoryOAuthLib.prototype.set_getOAuthRequestToken = function (oauth_token, oauth_secret) {
        this.return_oauth_token = oauth_token;
        this.return_oauth_secret = oauth_secret;
    };
    InMemoryOAuthLib.prototype.getOAuthRequestToken = function (callback) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.lastCalledMethod = 'getOAuthRequestToken';
                callback(this.err, this.return_oauth_token, this.return_oauth_secret, null);
                return [2 /*return*/];
            });
        });
    };
    InMemoryOAuthLib.prototype.set__performSecureRequest = function (oauth_token, oauth_secret, sessionHandle, sessionExpires) {
        this.returnAuthorisedToken = oauth_token;
        this.returnAuthorisedSecret = oauth_secret;
        this.returnSessionHandle = sessionHandle;
        this.oauth_expires_in = sessionExpires;
    };
    InMemoryOAuthLib.prototype._performSecureRequest = function (oauth_token, oauth_token_secret, verb, oauthAccessTokenPath, extraParams, something, something2, callback) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.lastCalledMethod = '_performSecureRequest';
                callback(this.err, "oauth_session_handle=" + this.returnSessionHandle + "&oauth_token_secret=" + this.returnAuthorisedSecret + "&oauth_token=" + this.returnAuthorisedToken + "&oauth_expires_in=" + this.oauth_expires_in);
                return [2 /*return*/];
            });
        });
    };
    InMemoryOAuthLib.prototype.getOAuthAccessToken = function (authedToken, authedSecret, oauthVerifier, callback) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.lastCalledMethod = 'getOAuthAccessToken';
                callback(this.err, this.returnAuthorisedToken, this.returnAuthorisedSecret, { oauth_session_handle: this.returnSessionHandle, oauth_expires_in: this.oauth_expires_in });
                return [2 /*return*/];
            });
        });
    };
    return InMemoryOAuthLib;
}());
exports.InMemoryOAuthLib = InMemoryOAuthLib;
