// xero-node/lib/internals/config-helper.js
"use strict";
/** @internalapi */
/** This second comment is required for typedoc to recognise the WHOLE FILE as @internalapi */
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("./utils");
var version = require('../../package.json').version;
/** @private */
function mapState(xeroConfig) {
    var cert = xeroConfig.privateKeyPath ? utils_1.getStringFromFile(xeroConfig.privateKeyPath) : null; // TODO don't read twice
    if (xeroConfig.privateKeyString) {
        cert = xeroConfig.privateKeyString;
    }
    if (xeroConfig.appType == 'private') {
        return {
            oauth_token: xeroConfig.consumerKey,
            oauth_token_secret: cert
        };
    }
    else if (xeroConfig.appType == 'public') {
        return null;
    }
    else if (xeroConfig.appType == 'partner') {
        return null;
    }
    else {
        throw new Error("Unrecognised app type: " + xeroConfig.appType + " (expected private|public|partner)");
    }
}
exports.mapState = mapState;
/** @private */
function mapConfig(xeroConfig, apiConfig) {
    // the logic for API_BASE can be used for testing against a mock server
    var API_BASE = process.env.XERO_API_BASE ? process.env.XERO_API_BASE : 'https://api.xero.com';
    var OAUTH_REQUEST_TOKEN_PATH = '/oauth/RequestToken';
    var OAUTH_ACCESS_TOKEN_PATH = '/oauth/AccessToken';
    var cert = xeroConfig.privateKeyPath ? utils_1.getStringFromFile(xeroConfig.privateKeyPath) : null;
    var userAgentString = 'NodeJS-XeroAPIClient';
    if (xeroConfig.userAgent) {
        userAgentString = userAgentString + '.' + xeroConfig.userAgent;
    }
    userAgentString = userAgentString + '.' + version + '.' + xeroConfig.consumerKey;
    if (xeroConfig.privateKeyString) {
        cert = xeroConfig.privateKeyString;
    }
    var oauthConfig = {
        apiBaseUrl: API_BASE,
        apiBasePath: apiConfig.apiBasePath || '',
        oauthRequestTokenPath: OAUTH_REQUEST_TOKEN_PATH,
        oauthAccessTokenPath: OAUTH_ACCESS_TOKEN_PATH,
        accept: 'application/json',
        userAgent: userAgentString,
        consumerKey: xeroConfig.consumerKey,
        consumerSecret: xeroConfig.consumerSecret,
        tenantType: apiConfig.tenantType || null,
        signatureMethod: undefined,
        callbackUrl: xeroConfig.callbackUrl ? xeroConfig.callbackUrl : null
    };
    if (xeroConfig.appType == 'private') {
        oauthConfig.consumerSecret = cert;
        oauthConfig.signatureMethod = 'RSA-SHA1';
    }
    else if (xeroConfig.appType == 'public') {
        oauthConfig.signatureMethod = 'HMAC-SHA1';
    }
    else if (xeroConfig.appType == 'partner') {
        oauthConfig.consumerSecret = cert;
        oauthConfig.signatureMethod = 'RSA-SHA1';
    }
    else {
        throw new Error("Unrecognised app type: " + xeroConfig.appType + " (expected private|public|partner)");
    }
    return oauthConfig;
}
exports.mapConfig = mapConfig;
