// xero-node/lib/internals/OAuth1HttpClient.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var http = require("http");
var https = require("https");
var oauth_1 = require("oauth");
var querystring = require("querystring");
var URL = require("url");
var XeroError_1 = require("../XeroError");
/** @private */
var OAuth1HttpClient = /** @class */ (function () {
    function OAuth1HttpClient(config, authState, oAuthLibFactory) {
        var _this = this;
        this.config = config;
        this.oAuthLibFactory = oAuthLibFactory;
        this._state = null;
        this.agent = null;
        this._defaultHeaders = {
            'Accept': 'application/json',
            'User-Agent': this.config.userAgent
        };
        this.getRequestToken = function () { return __awaiter(_this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                this.resetToDefaultHeaders();
                return [2 /*return*/, new Promise(function (resolve, reject) {
                        _this.oauthLib.getOAuthRequestToken(function (err, oauth_token, oauth_token_secret, result) {
                            if (err) {
                                reject(err.statusCode ? new XeroError_1.XeroError(err.statusCode, err.data, null) : err);
                            }
                            else {
                                resolve({
                                    oauth_token: oauth_token,
                                    oauth_token_secret: oauth_token_secret
                                });
                            }
                        });
                    })];
            });
        }); };
        this.buildAuthoriseUrl = function (requestToken) {
            return _this.config.apiBaseUrl + "/oauth/Authorize?oauth_token=" + requestToken.oauth_token;
        };
        this.swapRequestTokenforAccessToken = function (requestToken, oauth_verifier) { return __awaiter(_this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                this.resetToDefaultHeaders();
                return [2 /*return*/, new Promise(function (resolve, reject) {
                        _this.oauthLib.getOAuthAccessToken(requestToken.oauth_token, requestToken.oauth_token_secret, oauth_verifier, function (err, oauth_token, oauth_token_secret, results) {
                            if (err) {
                                reject(err.statusCode ? new XeroError_1.XeroError(err.statusCode, err.data, null) : err);
                            }
                            else {
                                var currentMilliseconds = new Date().getTime();
                                var expDate = new Date(currentMilliseconds + (results.oauth_expires_in * 1000));
                                var oauthState = {
                                    oauth_token: oauth_token,
                                    oauth_token_secret: oauth_token_secret,
                                    oauth_session_handle: results.oauth_session_handle,
                                    oauth_expires_at: expDate
                                };
                                _this.setState(oauthState);
                                resolve(oauthState);
                            }
                        });
                    })];
            });
        }); };
        this.refreshAccessToken = function () { return __awaiter(_this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, new Promise(function (resolve, reject) {
                        // We're accessing this "private" method as the lib does not allow refresh with oauth_session_handle.
                        _this.oauthLib._performSecureRequest(_this._state.oauth_token, _this._state.oauth_token_secret, 'POST', _this.config.apiBaseUrl + _this.config.oauthAccessTokenPath, { oauth_session_handle: _this._state.oauth_session_handle }, null, null, function (err, data, response) {
                            if (err) {
                                reject(err.statusCode ? new XeroError_1.XeroError(err.statusCode, err.data, response ? response.headers : null) : err);
                            }
                            else {
                                var results = querystring.parse(data);
                                var currentMilliseconds = new Date().getTime();
                                var expDate = new Date(currentMilliseconds + (results.oauth_expires_in * 1000));
                                var oauthState = {
                                    oauth_token: results.oauth_token,
                                    oauth_token_secret: results.oauth_token_secret,
                                    oauth_session_handle: results.oauth_session_handle,
                                    oauth_expires_at: expDate
                                };
                                _this.setState(oauthState);
                                resolve(oauthState);
                            }
                        });
                    })];
            });
        }); };
        this.writeUTF8ResponseToStream = function (endpoint, mimeType, writeStream) {
            _this.resetToDefaultHeaders();
            return new Promise(function (resolve, reject) {
                _this.assertAccessTokenIsSet();
                var oauthForPdf = _this.oAuthLibFactory(__assign({}, _this.config, { accept: mimeType }));
                var request = oauthForPdf.get(_this.config.apiBaseUrl + _this.config.apiBasePath + endpoint, _this._state.oauth_token, _this._state.oauth_token_secret);
                request.addListener('response', function (response) {
                    response.on('data', function (chunk) {
                        writeStream.write(chunk);
                    });
                    response.on('end', function () {
                        writeStream.end();
                        return resolve();
                    });
                    response.on('close', function () {
                        writeStream.end();
                        return resolve();
                    });
                });
                request.end();
            });
        };
        this.writeBinaryResponseToStream = function (endpoint, mimeType, writeStream) {
            _this.resetToDefaultHeaders();
            return new Promise(function (resolve, reject) {
                _this.assertAccessTokenIsSet();
                var forPDF = _this.oAuthLibFactory(__assign({}, _this.config, { accept: mimeType }));
                _this._OURperformSecureRequest(_this._state.oauth_token, _this._state.oauth_token_secret, 'GET', _this.config.apiBaseUrl + _this.config.apiBasePath + endpoint, function (err, data, httpResponse) {
                    // data is the body of the response
                    if (err) {
                        reject(err.statusCode ? new XeroError_1.XeroError(err.statusCode, err.data, httpResponse.headers) : err);
                    }
                    else {
                        var buffer = new Buffer(data, 'binary');
                        writeStream.write(buffer, function () {
                            writeStream.close();
                            return resolve();
                        });
                    }
                }, forPDF);
            });
        };
        this._OURperformSecureRequest = function (oauth_token, oauth_token_secret, method, url, callback, oauthForBinary) {
            // This code was copied out from the lib as it does not support binary downloads.
            var orderedParameters = oauthForBinary._prepareParameters(oauth_token, oauth_token_secret, method, url, null);
            var parsedUrl = URL.parse(url, false);
            if (parsedUrl.protocol == 'http:' && !parsedUrl.port) {
                parsedUrl.port = '80';
            }
            if (parsedUrl.protocol == 'https:' && !parsedUrl.port) {
                parsedUrl.port = '443';
            }
            var headers = {};
            var authorization = oauthForBinary._buildAuthorizationHeaders(orderedParameters);
            headers['Authorization'] = authorization;
            headers['Host'] = parsedUrl.host;
            for (var key in this._headers) {
                if (this._headers.hasOwnProperty(key)) {
                    headers[key] = this._headers[key];
                }
            }
            headers['Content-length'] = 0;
            var path;
            if (!parsedUrl.pathname || parsedUrl.pathname == '') {
                parsedUrl.pathname = '/';
            }
            // tslint:disable-next-line:prefer-conditional-expression
            if (parsedUrl.query) {
                path = parsedUrl.pathname + '?' + parsedUrl.query;
            }
            else {
                path = parsedUrl.pathname;
            }
            var request;
            // tslint:disable-next-line:prefer-conditional-expression
            if (parsedUrl.protocol == 'https:') {
                request = oauthForBinary._createClient(parsedUrl.port, parsedUrl.hostname, method, path, headers, true);
            }
            else {
                request = oauthForBinary._createClient(parsedUrl.port, parsedUrl.hostname, method, path, headers);
            }
            var data = '';
            function passBackControl(response) {
                if (response.statusCode >= 200 && response.statusCode <= 299) {
                    callback(null, data, response);
                }
                else {
                    callback({ statusCode: response.statusCode, data: data }, data, response);
                }
            }
            request.on('response', function (response) {
                response.setEncoding('binary');
                response.on('data', function (chunk) {
                    data += chunk;
                });
                response.on('end', function () {
                    passBackControl(response);
                });
                // response.on('close', function() {
                //     passBackControl(response);
                // });
            });
            request.on('error', function (err) {
                callback(err);
            });
            request.end();
        };
        this.readStreamToRequest = function (endpoint, mimeType, size, readStream) {
            _this.resetToDefaultHeaders();
            return new Promise(function (resolve, reject) {
                _this.assertAccessTokenIsSet();
                _this.resetToDefaultHeaders();
                var bufs = [];
                readStream
                    .on('data', function (chunk) {
                    bufs.push(chunk);
                })
                    .on('end', function () {
                    _this.oauthLib._headers = __assign({}, _this._defaultHeaders, { 'Content-Type': mimeType, 'Content-Length': size });
                    _this.oauthLib.post(_this.config.apiBaseUrl + _this.config.apiBasePath + endpoint, // url
                    _this._state.oauth_token, _this._state.oauth_token_secret, Buffer.concat(bufs), mimeType, function (err, data, httpResponse) {
                        if (err) {
                            reject(new XeroError_1.XeroError(httpResponse.statusCode, data, httpResponse ? httpResponse.headers : null));
                        }
                        else {
                            var toReturn = JSON.parse(data);
                            return resolve(toReturn);
                        }
                    });
                });
            });
        };
        this.get = function (endpoint, customHeaders) { return __awaiter(_this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                this.resetToDefaultHeaders();
                this.oauthLib._headers = __assign({}, this._defaultHeaders, customHeaders);
                return [2 /*return*/, new Promise(function (resolve, reject) {
                        _this.assertAccessTokenIsSet();
                        _this.oauthLib.get(_this.config.apiBaseUrl + _this.config.apiBasePath + endpoint, // url
                        _this._state.oauth_token, _this._state.oauth_token_secret, function (err, data, httpResponse) {
                            // data is the body of the response
                            if (err) {
                                reject(err.statusCode ? new XeroError_1.XeroError(err.statusCode, err.data, httpResponse ? httpResponse.headers : null) : err);
                            }
                            else {
                                var toReturn = JSON.parse(data);
                                return resolve(toReturn);
                            }
                        });
                    })];
            });
        }); };
        this.put = function (endpoint, body, customHeaders) { return __awaiter(_this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                this.resetToDefaultHeaders();
                this.oauthLib._headers = __assign({}, this._defaultHeaders, customHeaders);
                this.assertAccessTokenIsSet();
                return [2 /*return*/, new Promise(function (resolve, reject) {
                        _this.oauthLib.put(_this.config.apiBaseUrl + _this.config.apiBasePath + endpoint, // url
                        _this._state.oauth_token, _this._state.oauth_token_secret, JSON.stringify(body), // Had to do this not sure if there is another way
                        'application/json', function (err, data, httpResponse) {
                            // data is the body of the response
                            if (err) {
                                reject(err.statusCode ? new XeroError_1.XeroError(err.statusCode, err.data, httpResponse ? httpResponse.headers : null) : err);
                            }
                            else {
                                var toReturn = JSON.parse(data);
                                return resolve(toReturn);
                            }
                        });
                    })];
            });
        }); };
        this.post = function (endpoint, body, customHeaders) { return __awaiter(_this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                this.resetToDefaultHeaders();
                this.oauthLib._headers = __assign({}, this._defaultHeaders, customHeaders);
                this.assertAccessTokenIsSet();
                return [2 /*return*/, new Promise(function (resolve, reject) {
                        _this.oauthLib.post(_this.config.apiBaseUrl + _this.config.apiBasePath + endpoint, // url
                        _this._state.oauth_token, _this._state.oauth_token_secret, JSON.stringify(body), // Had to do this not sure if there is another way
                        'application/json', function (err, data, httpResponse) {
                            // data is the body of the response
                            if (err) {
                                reject(err.statusCode ? new XeroError_1.XeroError(err.statusCode, err.data, httpResponse ? httpResponse.headers : null) : err);
                            }
                            else {
                                var toReturn = null;
                                if (data) {
                                    toReturn = JSON.parse(data);
                                }
                                return resolve(toReturn);
                            }
                        });
                    })];
            });
        }); };
        this.delete = function (endpoint, customHeaders) { return __awaiter(_this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                this.resetToDefaultHeaders();
                this.oauthLib._headers = __assign({}, this._defaultHeaders, customHeaders);
                this.assertAccessTokenIsSet();
                return [2 /*return*/, new Promise(function (resolve, reject) {
                        _this.oauthLib.delete(_this.config.apiBaseUrl + _this.config.apiBasePath + endpoint, // url
                        _this._state.oauth_token, _this._state.oauth_token_secret, function (err, data, httpResponse) {
                            // data is the body of the response
                            if (err) {
                                reject(err.statusCode ? new XeroError_1.XeroError(err.statusCode, err.data, httpResponse ? httpResponse.headers : null) : err);
                            }
                            else {
                                var toReturn = null;
                                if (data) {
                                    toReturn = JSON.parse(data);
                                }
                                return resolve(toReturn);
                            }
                        });
                    })];
            });
        }); };
        if (authState) {
            this._state = authState;
        }
        if (!this.oAuthLibFactory) {
            this.oAuthLibFactory = function (passedInConfig) {
                var requestTokenPath = passedInConfig.oauthRequestTokenPath;
                if (passedInConfig.tenantType) {
                    requestTokenPath += "?tenantType=" + passedInConfig.tenantType;
                }
                return new oauth_1.OAuth(passedInConfig.apiBaseUrl + requestTokenPath, // requestTokenUrl
                passedInConfig.apiBaseUrl + passedInConfig.oauthAccessTokenPath, // accessTokenUrl
                passedInConfig.consumerKey, // consumerKey
                passedInConfig.consumerSecret, // consumerSecret
                '1.0A', // version
                config.callbackUrl, // authorize_callback
                passedInConfig.signatureMethod, // signatureMethod. Neesds to ve "RSA-SHA1" for Private. "HMAC-SHA1" for public
                null, // nonceSize
                {
                    'Accept': passedInConfig.accept,
                    'User-Agent': passedInConfig.userAgent
                });
            };
        }
        this.oauthLib = this.oAuthLibFactory(this.config);
        this.oauthLib._createClient = this._createHttpClientWithProxySupport.bind(this);
    }
    OAuth1HttpClient.prototype.resetToDefaultHeaders = function () {
        this.oauthLib._headers = this._defaultHeaders;
    };
    OAuth1HttpClient.prototype.setState = function (newState) {
        this._state = __assign({}, this._state, newState);
    };
    OAuth1HttpClient.prototype.assertAccessTokenIsSet = function () {
        if (!this._state.oauth_token) {
            throw new Error('Missing access token. Acquire a new access token by following the oauth flow or call setState() to use an existing token.');
        }
    };
    // Monkey-patched OAuthLib _createClient function to add proxy support
    OAuth1HttpClient.prototype._createHttpClientWithProxySupport = function (port, hostname, method, path, headers, sslEnabled) {
        var options = {
            host: hostname,
            port: port,
            path: path,
            method: method,
            headers: headers
        };
        var httpModel = sslEnabled ? https : http;
        if (this.agent) {
            options.agent = this.agent;
        }
        return httpModel.request(options);
    };
    return OAuth1HttpClient;
}());
exports.OAuth1HttpClient = OAuth1HttpClient;
